### PR TITLE
`Development`: Validate postal code and final grade input

### PR DESCRIPTION
**Pull request by Douglas-Sarah**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://confluence.aet.cit.tum.de/spaces/AP/pages/257786006/Language+Guidelines+Client).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311714/PR+Guidelines).

#### Client

- [x] I **strictly** followed the [client coding and design guidelines](https://confluence.aet.cit.tum.de/spaces/AP/pages/256311716/Client+Guidelines).
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Currently, the application form accepts any postal code input without validation, which can lead to invalid data being submitted. Users may enter postal codes that don't match the format requirements for their selected country, causing issues downstream. Additionally, the grade input fields on page 2 were using string inputs when numeric inputs would provide better UX and validation.

### Description

This PR implements comprehensive postal code validation and improves form input handling:

#### Postal Code Validation
- **Added `postal-codes-js` library** for country-specific postal code validation
- **Implemented `postalCodeValidator`** that validates postal codes against the selected country's format
- **Updated country field** from string to `SelectOption` type for better type safety
- **Added comprehensive country list** with ISO country codes in `nationalities.ts`
- **Added validation error messages** in both English and German translations

#### Form Improvements
- **Enhanced Page 2 grade inputs** to use `NumberInputComponent` instead of string inputs for bachelor grades
- **Added conditional rendering** based on grading scale (currently supports 1-4 scale)
- **Improved type safety** with proper grade conversion methods

#### Technical Changes
- Updated `ApplicationCreationPage1Data` interface to use `SelectOption` for country
- Modified `getPage1FromApplication` to properly map country data
- Added postal code validator with translation key support
- Updated test mocks to reflect new country structure
- Added helper methods for grade number conversion

#### Postal Code Validation Testing
0. Login as applicant
1. **Navigate to Application Page 1**
2. **Select different countries** from the country dropdown
3. **Enter invalid postal codes** for each country:
   - Germany (DE): Try "ABC123" (should show validation error)
   - United States (US): Try "12345-ABCD" (should show validation error)
   - United Kingdom (GB): Try "12345" (should show validation error)
4. **Enter valid postal codes**:
   - Germany: "10115" or "80331"
   - United States: "90210" or "10001"
   - United Kingdom: "SW1A 1AA" or "M1 1AA"
5. **Verify validation messages** appear in correct language (EN/DE)
6. **Test edge cases**:
   - Leave postal code empty with country selected
   - Select country after entering postal code
   - Change country with existing postal code

#### Grade Input Testing
1. **Navigate to Application Page 2**
2. **Select "1-4 Scale" for bachelor grading scale**
3. **Verify number input appears** with proper min/max constraints (1.0 - 4.0)
4. **Test input validation**:
   - Enter values outside range (should be constrained)
   - Enter decimal values (should accept 1 decimal place)
   - Leave field empty and verify required validation

### Regression Testing
1. **Complete full form flow** from page 1 to submission
2. **Verify existing functionality** still works:
   - All other form fields
   - Form navigation
   - Data persistence between pages
3. **Test form pre-population** when editing existing applications
4. **Verify translations** work correctly in both languages

#### Code Review

- [x] Code Review 1

#### Manual Tests

- [x] Test 1

### Screenshots

image
image

### CODE REVIEW

**Review by Bofan-Zhu — CHANGES_REQUESTED** [review]
Screenshot 2025-08-24 at 12 06 49
The input for the grade feels a bit buggy and confusing. When I remove my input, the field shows NaN and when I remove the fractional digit (e.g., 9 from 2.9) the number immediately shows 2.0.
The tooltip of both forms doesn't work anymore (If you look at the picture it also slightly looks different from the other tooltips)

Validation for postal code lgtm 👍. Just to double check, the issue did not include to also check for correct postal codes right (Validate if the postal code actually exists)?

**Review by Bofan-Zhu — APPROVED** [review]
LGTM 👍

File: src/main/webapp/app/application/application-creation/application-creation-page1/application-creation-page1.component.ts [hunk 1]
```diff
@@ -84,19 +95,21 @@ export default class ApplicationCreationPage1Component {
...
       lastName: [currentData.lastName, Validators.required],
-      email: [currentData.email, Validators.required],
+      email: [currentData.email, [Validators.required, Validators.email]],
       phoneNumber: [currentData.phoneNumber, Validators.required],
 
       street: [currentData.street, Validators.required],
       city: [currentData.city, Validators.required],
       country: [currentData.country, Validators.required],
-      postcode: [currentData.postcode, Validators.required],
+      postcode: [currentData.postcode, [Validators.required, postalCodeValidator(() => this.data().country?.value as string)]],
```
[the remark this answers is not in the document]
Comment (Douglas-Sarah) [reply src/main/webapp/app/application/application-creation/application-creation-page1/application-creation-page1.component.ts:112]: The problem is that the SelectComponent has the following property:
```
export type SelectOption = {
  name: string;
  value: string | number;
  icon?: string;
};
...
export class SelectComponent {
  items = input<SelectOption[]>([]);
  selected = input<SelectOption | undefined>(undefined);
```

This means, everywhere where SelectComponent is used, its value is being casted.

I would propose to change SelectComponent to make it generic:

```
export type SelectOption<T> = {
  name: string;
  value: T;
  icon?: string;
};
...
export class SelectComponent<T> {
  items = input<SelectOption<T>[]>([]);
  selected = input<SelectOption<T> | undefined>(undefined);
```
This is allowed in the underlying primeng component, to give ObjectTypes as values, and we would gain typesafety.

But as this PR is quite long already, I should change this in a separate PR, right?

File: src/main/webapp/app/application/application-creation/application-creation-page1/application-creation-page1.component.ts [hunk 2]
```diff
@@ -54,11 +55,21 @@ export const getPage1FromApplication = (application: ApplicationForApplicantDTO)
...
   };
 };
 
+function postalCodeValidator(getCountryFn: () => string | undefined): ValidatorFn {
+  return (control: AbstractControl): ValidationErrors | null => {
+    const country = getCountryFn()?.toUpperCase();
+    const value = control.value;
+    if (!country || !value) return null;
+    const result = postalCodes.validate(country, value);
+    return result === true ? null : { invalidPostalCode: 'entity.applicationPage1.validation.postalCode' };
```
[the remark this answers is not in the document]
Comment (Douglas-Sarah) [reply src/main/webapp/app/application/application-creation/application-creation-page1/application-creation-page1.component.ts:69]: Renamed 👍

File: src/main/webapp/app/application/application-creation/application-creation-page2/application-creation-page2.component.ts [hunk 3]
```diff
@@ -115,4 +119,25 @@ export default class ApplicationCreationPage2Component {
...
   }
+
+  getBachelorGradeAsNumber(): number | undefined {
+    const bachelorGrade = this.data().bachelorGrade;
+    return bachelorGrade === '' ? undefined : Number.parseFloat(bachelorGrade);
+  }
+  setBachelorGradeAsNumber(ev: number | undefined): void {
+    this.data.set({
+      ...this.data(),
+      bachelorGrade: ev ? ev.toString() : '',
```
[the remark this answers is not in the document]
Comment (Douglas-Sarah) [reply src/main/webapp/app/application/application-creation/application-creation-page2/application-creation-page2.component.ts:130]: Fixed, thanks for catching that 👍

File: src/main/webapp/app/application/application-creation/application-creation-page2/application-creation-page2.component.ts [hunk 4]
```diff
@@ -115,4 +119,25 @@ export default class ApplicationCreationPage2Component {
...
+  }
+  getMasterGradeAsNumber(): number | undefined {
+    const masterGrade = this.data().masterGrade;
+    return masterGrade === '' ? undefined : Number.parseFloat(masterGrade);
+  }
+  setMasterGradeAsNumber(ev: number | undefined): void {
+    this.data.set({
+      ...this.data(),
+      masterGrade: ev ? ev.toString() : undefined,
+    } as ApplicationCreationPage2Data);
```
[the remark this answers is not in the document]
Comment (Douglas-Sarah) [reply src/main/webapp/app/application/application-creation/application-creation-page2/application-creation-page2.component.ts:141]: Removed the casting 👍

File: src/main/webapp/app/application/application-creation/application-creation-page1/application-creation-page1.component.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rabeatwork) [reply src/main/webapp/app/application/application-creation/application-creation-page1/application-creation-page1.component.ts:112]: Yes, let's refactor it in a separate PR.

**Review by rabeatwork — APPROVED** [review]
